From 6c42305c46f6f4e68d5bc2cf8b2dd02bcea8c639 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:00:17 -0400
Subject: [ethernet] mlx4: Handle vxlan steering rules for mac address changes

Message-id: <ee7f1001997a8eb6ce334d4f0f6ab65b9d9b8b01.1410425016.git.avadai@redhat.com>
Patchwork-id: 93617
O-Subject: [Patch RHEL7.1 V1 081/197] net/mlx4_en: Handle vxlan steering rules for mac address changes
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Or Gerlitz <ogerlitz@mellanox.com>

When the device mac address is changed, we must deregister the vxlan
steering rule associated with the previous mac, and register a new
steering rule using the new mac.

Signed-off-by: Or Gerlitz <ogerlitz@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 2a2083f7f3568c0192daa6ac0e6fa35d953f47bd)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_netdev.c b/drivers/net/ethernet/mellanox/mlx4/en_netdev.c
index 3db5946..b4f6547 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_netdev.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_netdev.c
@@ -742,6 +742,14 @@ static int mlx4_en_replace_mac(struct mlx4_en_priv *priv, int qpn,
     err = mlx4_en_uc_steer_add(priv, new_mac,
           &qpn,
           &entry->reg_id);
+    if (err)
+     return err;
+    if (priv->tunnel_reg_id) {
+     mlx4_flow_detach(priv->mdev->dev, priv->tunnel_reg_id);
+     priv->tunnel_reg_id = 0;
+    }
+    err = mlx4_en_tunnel_steer_add(priv, new_mac, qpn,
+              &priv->tunnel_reg_id);
     return err;
    }
   }
-- 
1.7.1